### PR TITLE
Introduce into_raw_non_null on Rc and Arc

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: src/liballoc/rc.rs [hunk 1]
```diff
@@ -435,6 +435,26 @@ impl<T: ?Sized> Rc<T> {
...
+    ///
+    /// let x = Rc::new(10);
+    /// let ptr = Rc::into_non_null(x);
+    /// let deref = unsafe { *ptr.as_ref() };
+    /// assert_eq!(deref, 10);
+    /// ```
+    #[unstable(feature = "rc_into_non_null", issue = "0")]
+    #[inline]
+    pub fn into_non_null(this: Self) -> NonNull<T> {
+        unsafe { NonNull::new_unchecked(Rc::into_raw(this) as *mut _) }
```
Comment (Centril) [inline src/liballoc/rc.rs]: A few words re. why this is safe would be good.

File: src/liballoc/sync.rs [hunk 2]
```diff
@@ -413,6 +413,26 @@ impl<T: ?Sized> Arc<T> {
...
+    ///
+    /// let x = Arc::new(10);
+    /// let ptr = Arc::into_non_null(x);
+    /// let deref = unsafe { *ptr.as_ref() };
+    /// assert_eq!(deref, 10);
+    /// ```
+    #[unstable(feature = "rc_into_non_null", issue = "0")]
+    #[inline]
+    pub fn into_non_null(this: Self) -> NonNull<T> {
+        unsafe { NonNull::new_unchecked(Arc::into_raw(this) as *mut _) }
```
Comment (Centril) [inline src/liballoc/sync.rs]: and here...

**Review by Centril — DISMISSED** [review]

**Comment by alexcrichton** [comment]
Thanks for this! On `Box` this is called `into_raw_non_null`, so should the name be updated to match that?

**Comment by dwijnand** [comment]
> On `Box` this is called `into_raw_non_null`, so should the name be updated to match that?

I assumed the "raw" in Box's version is because it already had an `into_raw` method, but I wasn't sure if it was appropriate for Arc/Rc. I also don't know if the rename is worth it, but I'm happy to make the old name conform with the new names or make the new names conform with the old name. Do you have a preference?

**Comment by alexcrichton** [comment]
Nah yeah I think we'll want to prioritize consistency across pointer types, and I think that `{Arc,Rc}::into_raw` should exist as well!

**Comment by dwijnand** [comment]
> and I think that `{Arc,Rc}::into_raw` should exist as well!

I'm blind: I'm _using_ `{Arc,Rc}::into_raw` in the implementation of `into_raw_non_null` 🙄.

All done!

**Comment by alexcrichton** [comment]
As one final nit, could the tracking issue be pointed to the same one as for Box?

**Comment by dwijnand** [comment]
Sure!

**Pull request merged**